File: debugLoglik/R2/CreateModel.R
# Function creating the pomp model

source("ModelSnippet.R")

create_pomp_model <- function(data, covars, t0, 
                              dt=0.005, params.notransform=NULL) {
  dt <- dt
  covars <- subset(covars, select = c("time", "temperature"))

  trans1 <- c("p","omega","delta",
              "mu_e","mu_ql","mu_el","mu_qn","mu_en","mu_qa","mu_ea","mu_h",
              "beta_nh","gamma",
              "lambda_l","lambda_n","lambda_a","alpha","kappa","Tf","T_min_l",
              paste0(statenames[1:11],"0"))
  trans2 <- c("f_l","f_n","f_a","c")
  trans1 <- trans1[which(!(trans1 %in% params.notransform))]
  trans2 <- trans2[which(!(trans2 %in% params.notransform))]
  
  

  # Create the pomp model
  TBE <- pomp(
    data = subset(data, select = c("time", "reports")),
    times = "time",
    t0 = t0,
    covar = covars,
    tcovar = "time",
    zeronames = c("cases"),
    paramnames = c("p","omega","delta",
                   "mu_e","mu_ql","mu_el","mu_qn","mu_en","mu_qa","mu_ea","mu_h",
                   "beta_nh","beta_hl","beta_hn","gamma",
                   "obsprob", 
                   "lambda_l","lambda_n","lambda_a","alpha","f_l","f_n","f_a","kappa", "Tf", "c","T_min_l",
                   paste0(statenames,"0")),
    dmeasure = dObs,
    rmeasure = rObs,
    skeleton = vectorfield(skel),
    log.trans = trans1,
    logit.trans = trans2,
    fromEstimationScale  = function(params, log.trans, logit.trans,  ...){
      params[log.trans] <- exp(params[log.trans])
      params[logit.trans] <- plogis(params[logit.trans])
      params[c("H_s0","H_i0")] <- exp(params[c("H_s0","H_i0")]) / (1 + sum(exp(params[c("H_s0","H_i0")])))
      params
    },
    toEstimationScale = function(params, log.trans, logit.trans, ...){
      params[log.trans] <- log(params[log.trans])
      params[logit.trans] <- qlogis(params[logit.trans])
      params[c("H_s0","H_i0")] <- log(params[c("H_s0","H_i0")] / (1 - sum(params[c("H_s0","H_i0")])))
      params
    },
    initializer = rInit,
    statenames=c(statenames, "cases")
  )

  output <- list(model=TBE)
  print("Model created.")
  return(output)
}

File: debugLoglik/js/createModel.js

/**
 *  @file        CreateModel.js
 *               Transforms parameters to and from the scale. 
 *                 
 *  @autor       Nazila Akhavan, nazila@kingsds.network
 *  @date        July 2019
 */
snippet = require("./modelSnippet.js")
mathLib = require ("./mathLib.js")
Index = require('./indices')

let createModel = {}
createModel.createPompModel = function(data, covars, t0, dt = 0.005, params_notransform = null) { 

  let statenames = snippet.statenames()
  trans1 = [Index.p,Index.omega,Index.delta, Index.mu_e,Index.mu_ql, Index.mu_el,
              Index.mu_qn,Index.mu_en,Index.mu_qa,Index.mu_ea,Index.mu_h, Index.beta_nh,
              Index.lambda_l,Index.lambda_n,Index.lambda_a,Index.alpha, Index.kappa, Index.Tf, Index.T_min_l, Index.gamma]
  for ( let i = 0; i < 11; i++) {
    trans1.push(statenames[i])
  }
  
  // Remove parameters in trans1 that exist in params_notransform.
  for ( let i = 0; i < trans1.length; i++) {
    for ( let j = 0; j < params_notransform.length; j++) {
      if ( trans1[i] === params_notransform[j]) {
        trans1.splice(i,1)
        i -= 1
      }
    }
  }

  trans2 = [Index.f_l,Index.f_n,Index.f_a,Index.c]
  // Remove parameters in trans2 that exist in params_notransform.
  for ( let i = 0; i < trans2.length; i++) {
    for ( let j = 0; j < params_notransform.length; j++) {
      if ( trans2[i] === params_notransform[j]) {
        trans2.splice(i,1)
        i -= 1
      }
    }
  }

  // Trans1 uses log and trans2 uses logit 
  return [trans1, trans2]

}  
createModel.fromEstimationScale = function(params, logTrans, logitTrans){
  for ( let i = 0; i < logTrans.length; i++) {
    params[logTrans [i]] = Math.exp(params[logTrans[i]])
  }
  for ( let i = 0; i < logitTrans.length; i++) {
    params[logitTrans [i]] = mathLib.plogis(params[logitTrans[i]])
  }      
  let sumH = Math.exp(params[Index.H_s]) + Math.exp(params[Index.H_i])
  params[Index.H_s] = Math.exp(params[Index.H_s]) / (1 + sumH)
  params[Index.H_i] = Math.exp(params[Index.H_i]) / (1 + sumH)
  return params
} 

createModel.toEstimationScale = function(params, logTrans, logitTrans){
  for ( let i = 0; i < logTrans.length; i++) {
    params[logTrans [i]] = Math.log(params[logTrans[i]])
  }
  for ( let i = 0; i < logitTrans.length; i++) {
    params[logitTrans [i]] = mathLib.qlogis(params[logitTrans[i]])
  }      
  let sumH = params[Index.H_s] + params[Index.H_i]
  params[Index.H_s] = Math.log(params[Index.H_s] / (1 - sumH))
  params[Index.H_i] = Math.log(params[Index.H_i] / (1 - sumH))
  return params
}   


  module.exports = createModel